packages/flutter/lib/src/material/dropdown_menu_form_field.dart

```diff
@@ -0,0 +1,146 @@
+    required List<DropdownMenuEntry<T>> dropdownMenuEntries,
+    List<TextInputFormatter>? inputFormatters,
+    DropdownMenuCloseBehavior closeBehavior = DropdownMenuCloseBehavior.all,
+    super.onSaved,
+    AutovalidateMode? autovalidateMode,
+    super.validator,
+    super.forceErrorText,
+  }) : super(
+         initialValue: initialSelection,
+         autovalidateMode: autovalidateMode ?? AutovalidateMode.disabled,
+         builder: (FormFieldState<T> field) {
+           final _DropdownMenuFormFieldState<T> state = field as _DropdownMenuFormFieldState<T>;
+           void onSelectedHandler(T? value) {
+             field.didChange(value);
+             onSelected?.call(value);
+           }
+
+           return DropdownMenu<T>(
+             enabled: enabled,
+             width: width,
+             menuHeight: menuHeight,
+             leadingIcon: leadingIcon,
+             trailingIcon: trailingIcon,
+             label: label,
+             hintText: hintText,
+             helperText: helperText,
+             errorText: state.errorText,
+             selectedTrailingIcon: selectedTrailingIcon,
+             enableFilter: enableFilter,
+             enableSearch: enableSearch,
+             textStyle: textStyle,
+             textAlign: textAlign,
+             inputDecorationTheme: inputDecorationTheme,
+             menuStyle: menuStyle,
+             controller: controller,
+             initialSelection: state.value,
+             onSelected: onSelectedHandler,
+             requestFocusOnTap: requestFocusOnTap,
+             expandedInsets: expandedInsets,
+             alignmentOffset: alignmentOffset,
+             filterCallback: filterCallback,
+             searchCallback: searchCallback,
+             inputFormatters: inputFormatters,
+             closeBehavior: closeBehavior,
+             dropdownMenuEntries: dropdownMenuEntries,
+           );
+         },
+       );
+
+  /// The callback is called when a selection is made.
+  ///
+  /// Defaults to null. If null, only the text field is updated.
+  final ValueChanged<T?>? onSelected;
+
+  /// Controls the text being edited.
+  ///
+  /// If null, this widget will create its own [TextEditingController].
+  final TextEditingController? controller;
+
+  @override
+  FormFieldState<T> createState() => _DropdownMenuFormFieldState<T>();
+}
+
+class _DropdownMenuFormFieldState<T> extends FormFieldState<T> {
+  DropdownMenuFormField<T> get _dropdownMenuFormField => widget as DropdownMenuFormField<T>;
+
+  @override
+  void didChange(T? value) {
+    super.didChange(value);
+    _dropdownMenuFormField.onSelected?.call(value);
+  }
+
+  @override
+  void didUpdateWidget(DropdownMenuFormField<T> oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.initialValue != widget.initialValue) {
+      setValue(widget.initialValue);
+    }
+  }
+
+  @override
+  void reset() {
+    super.reset();
+    _dropdownMenuFormField.onSelected?.call(value);
+  }
+}
```

Comment:
Would it be useful to also override `restoreState`, like TextFormField does? https://github.com/flutter/flutter/blob/d3571a28a191b5bf2ae6111c4542541d14e6cb99/packages/flutter/lib/src/material/text_form_field.dart#L344

Reply:
I looked at this, the main problem is that Restoration API works for basic types (String, int, etc) but not for any subclass of Object (until serialization code is provided). And the selected value for a `DropdownMenu` could be of any type.

I will try to use the restoration API just for the inner text field (more precisely for its text controller) and in restoreState when a value is restored it might possible to check if it matches one of the dropdown menu entries.

Reply:
Ah true. Let me know how it goes, it would be nice to have if possible. Also add a test for it if you get it working.